patches/server/0977-Fix-multiple-spectator-bugs.patch

```diff
@@ -0,0 +1,61 @@
++++ b/src/main/java/net/minecraft/server/level/ChunkMap.java
+@@ -1535,6 +1535,7 @@ public class ChunkMap extends ChunkStorage implements ChunkHolder.PlayerProvider
+                 double d1 = vec3d_dx * vec3d_dx + vec3d_dz * vec3d_dz; // Paper
+                 double d2 = d0 * d0;
+                 boolean flag = d1 <= d2 && this.entity.broadcastToPlayer(player);
++                flag |= player.getCamera() == this.entity; // Paper - Make entities always visible for spectators, even when teleported far away (MC-107113)
+ 
+                 // CraftBukkit start - respect vanish API
+                 if (!player.getBukkitEntity().canSee(this.entity.getBukkitEntity())) {
+diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
+index 9d46536f80b5b3e6641fd377c02166a431edfd77..a6b55a8a527816230f2d1853c023161c992ca1da 100644
+--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
++++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
+@@ -11,6 +11,7 @@ import com.mojang.serialization.Dynamic;
+ import java.net.InetSocketAddress;
+ import java.net.SocketAddress;
+ import java.util.Collection;
++import java.util.Collections;
+ import java.util.Iterator;
+ import java.util.List;
+ import java.util.Objects;
+@@ -732,8 +733,25 @@ public class ServerPlayer extends Player {
+ 
+         if (entity != this) {
+             if (entity.isAlive()) {
+-                this.absMoveTo(entity.getX(), entity.getY(), entity.getZ(), entity.getYRot(), entity.getXRot());
+-                this.serverLevel().getChunkSource().move(this);
++                // Paper start - Fix spectator on cross-world teleports (MC-261799)
++                if (entity.level() != this.level()) {
++                    // Teleport ourselves to our camera
++                    this.getBukkitEntity().teleport(entity.getBukkitEntity().getLocation(), TeleportCause.SPECTATE);
++                    // Update the tracker of the other dimension for our cross-dimension teleport
++                    ChunkMap.TrackedEntity tracker = ((ServerLevel) entity.level()).getChunkSource().chunkMap.entityMap.get(entity.getId());
++                    if (tracker != null) {
++                        tracker.updatePlayer(this);
++                    }
++                    // Advise the client to start spectating again
++                    this.connection.send(new ClientboundSetCameraPacket(entity));
++                }else {
++                    // Paper: We send the player an additional teleport packet here to indicate that the position of itself has been moved.
++                    // Without this packet, if a player travels a too far distance, chunks will start to become invisible for our spectator. (MC-148993)
++                    this.connection.internalTeleport(entity.getX(), entity.getY(), entity.getZ(), entity.getYRot(), entity.getXRot(), Collections.emptySet());
```

Comment:
This really spam, i think this should be fixed in the client but if maintainer are ok, at least check if the player has moved in another chunk before sending another packet

Reply:
This could theoretically be adapted to only fire when walking over a chunk border, but can leave visual asymmetry of the world if the player has a relatively small render distance set.